### ISSUE TITLE
Sample of dataset or explanation

### ISSUE THREAD

**Issue by wheezy1749** [ladaapp/lada#19]
Can you share a sample of how you labeled the dataset or explain how the mask regions were labeled?

Curious how the NSFW model was trained to start. Since its used to create other datasets I'm sure. It seems to only care about NSFW regions and treats them all equally so I'm guessing these were all masked as the same object when close together regardless of how many there were.

If an SFW object was partially obstructing or overlapping the NSFW region was it included or excluded from the segment label? Or was part of the SFW (fingers for example) partially part of the mask.

Also, how many images were in the initial dataset? Thanks.

**Comment by wheezy1749** [comment]
Kinda answered my own question playing around with the detection model. Seems to have been trained to create a poly mask around each NSFW object (or multiple ones) as a mosaic would be. And as a mosaic would be. It treats any SFW object in that region as part of the model. Which makes sense for the use case.

I'd ask if you think a model that would better differentiate and segment (looking at sam2) precise parts of commonly mosaic'd objects would be worth the time to create.

Since we only use the NSFW detection to create a dataset for training the mosaic model are losing information by simply saying that the mosaic "block" should be treated equally in all cases. If the model was trained on different "parts" of what is likely below the mosaic it might better be able to re-construct the different parts more accurately. Just a thought. Would like to know if you've messed around with this idea.

**Comment by ladaapp** [comment]
That's right. I've labeled SFW regions (like a hand) that partially obstruct NSFW regions also as NSFW. That behavior matches most material I've watched. 
But in some newer videos I've seen those things are not being pixelated. Guess it depends on the tools/editor they are using. For older material were this censoring process was a fully manual process the mosaics masks were sometimes all over the place (very tight or very large). I assume on recent video they also have at least some if not full automation but I was not able to find some technical details how that works or what software is used.

I'd assume the current NSFW detection model would benefit from a bigger dataset to be more accurate. But labeling these images is taking a lot of time and I've spent enough weekends to draw these masks. I have not tested if differentiating NSFW into more specific categories as you suggested would help detection. I think labeling would need to be a lot more accurate than the current labeling though when annotating scenes where male/female parts are "very close to each other". But I'm not sure at this point which subcategories it should be labeled with. Obvious choices would be male/female gen.
Other thing that could be useful to classify mosaic size. For very huge mosaic blocks a custom mosaic removal modal which is deeper and more downscale/updscale layers could help.

But I've trained a custom mosaic restoration model based on a dataset consisting only of BJ POV videos. This did not result in noticeable better restoration quality. But when comparing it against the generic model these BJ-POV are also the easiest scenes with the best restoration quality. Would be interesting to train it on scenes the generic model is currently struggling...

Just some ideas though. I think it would be great to first study where exactly the mosaic restoration model is currently struggling (certain scenes, resolutions, mosaic sizes, quality/blur, etc). I cannot say for sure what is the issue with the current model at this point.
Anecdote: I think it is not specific scenes/objects. I've noticed that restoration quality of a certain JAV producer/series always results in relatively bad quality while with another JAV producer/series it get way better results. Mosaic size seems to be somewhat the same, not sure whats the crucial difference between these two.

Here are some stats about the three datasets the current models were trained on:

NSFW detection
3700 images

Mosaic detection
10200 images

Mosaic removal
23000 scenes/clips from 21000 files

Side note: I'm getting annoyed by not being able to decide which is the best terminology to describe what we're talking about here. Is it mosaic/pixel/censoring removal or thinking more about the original video restoration, recovery, reconstruction? :shrug:  Thoughts?

**Comment by wheezy1749** [comment]
Thanks. This info helps. Messing around with different scenarios. The cases where the model seems do to poorly are usually with female only regions. I'm wondering if the model had better knowledge of the types of objects behind the mosaic it could better infer from the surroundings to restore it better in cases where only one type of object is present. Or when SFW objects are behind the mosaic.

**Comment by ladaapp** [comment]
small update after training the mosaic restoration model with a bigger dataset and adjusted video degradations for the new model weights shipped in v0.6.0.
I increased the number of clips to 60k. As I haven't found a robust metric to describe restoration quality performance https://github.com/ladaapp/lada/issues/25 here are some anecdotes after evaluating some test files.

increasing the dataset size didn't help much in reducing the number of artifacts but still a small improvement was visible. deepmosaics is worse in general restoration quality but in comparison it does not show those artifacts. I think these are introduced by 2nd stage GAN training. One could validate this by training the model a long time on just stage 1 / pixel loss only and see if these artifacts show up there as well.

Besides just blowing up the dataset size with more videos I also did not filter out very low-quality content this time. Didn't notice much of an effect either way. I would have expected some outcome on the blurry-ness of the restored clips. Maybe there is some but didn't notice anything interesting in the samples I tested with.

In general artifacts not only appear on female only regions. I would say it mostly depends on ratio of mosaic block size vs object detail. For example, you can see those issues also if the camera is far away / 3rd person shots. For some videos the male part was covered by just two mosaic blocks in width then the restoration will just not do a good job. A similar scene where a less aggressive block size was used like 5 blocks to cover the thing  it works just fine. You can notice this also on fingers and fingernails which are even smaller in a bj scene. The male part is bigger and gets restored just fine but fingernails can struggle to be restored if block size is too large.

Another impact is the amount of change in the scene. if there is a lot of change even with a higher block size the model is better at restoration as it can get more data from previous and later frames which include change information of the mosaic blocks and therefore the underlying objects. You can see better restoration results for the same block size and the same object if there is much change compared to when the object is mostly still and not much changes during the scene.

The issue where mosaics are not restored at all even though they are detected by the mosaic detection model https://github.com/ladaapp/lada/issues/23 was not related to amount of type of scenes but just the degradation settings used for setting up the dataset. It still isn't perfect but a lot better now. My intuition would be that pixelization to create mosaics plus multiple rounds of video encoding should be enough to create a real-world dataset. I played a little bit with adding noise, blur, tiny rotations etc. but as far as I can tell this should not be necessary. I could not find any resources of how this censoring / mosaicing process works in real-life by the producers but I think its just original movie plus pixelation filter with some feathering/blur on the edges and video encoding to create the final file. then based on how the file is distributed there could be multiple rounds of reencoding/rescaling along the way that would reduce the quality of the mosaic.

it could be though that additional degradations/distortions help the model to learn better / less blurry /less artifacts even though those distortions are not present in real-life footage. I didn't do enough testing to confirm this.

If the mosaic block size is small enough and the video quality not garbage it is close to being perfect regardless of the object or scene arrangement. I Tested it on some films produced in the early Full HD days of maybe 2010 which had pretty moderate mosaic block sizes and I did not see any artifact in ca 6 hours of content.

Not really sure how to achieve this for more strongly pixelated content. For the current model maybe would make sense to add another down/upsampling layer and make it a bit deeper but cannot really tell, don't really understand enough honestly. https://github.com/ladaapp/lada/issues/7

Would be interested to know if you can confirm any of this or if you found some other insights
